snewpdag/plugins/DtsCalculator.py

```diff
@@ -0,0 +1,117 @@
+"""
+DtsCalculator: compute neutrino time dts between each detector sending an alert.
+
+Input payload:
+    data[neutrino_time] = observed burst time for a detector (s,ns). For mc trials use
+    the DetectorTime.py plugin to generate this.
+
+Output payload:
+    dts: a dictionary of time differences. Keys are of form (det1, det2). The values are themselves dictionaries with
+    the following keys:
+    dt: time tuple (s,ns) of t1-t2
+    t1: neutrino time tuple (s,ns) of det1
+    t2: neutrino time tuple (s,ns) of det2
+"""
+
+import logging
+import numpy as np
+from snewpdag.dag import Node, lib
+
+class DtsCalculator(Node):
+        def __init__(self, **kwargs):
+            self.valid = [False] * 10  # flags indicating valid data from sources
+            self.t = [None] * 10  # neutrino time for each detector
+            self.dets_names = [None] * 10
+            self.dets_nu_times = [0.0] * 10
+            self.map = {}
+            # self.uncertainties = [0.0] * 10  # time uncertainties for each source
+            # if kwargs.pop('detector_location_file'):
+            #     self.db = DetectorDB(kwargs.pop('detector_location_file'))
+            super().__init__(**kwargs)
+
+        def find_nu_times(self, data):
+            '''
+            Extract neutrino time for each detector
+            '''
+            # extract neutrino time for each detector:
+            index = self.last_watch_index()
+            burst_time = self.t[index]
+            self.t[index] = data.get('neutrino_time')
+            for det_name, nu_time in data['gen']['neutrino_times'].items():
+                if nu_time == self.t[index]:
+                    # det = self.db.get(det_name)
+                    # det_uncertainty = det.sigma
+                    # self.uncertainties[index] = det_uncertainty
+                    self.dets_names[index] = det_name
+                    self.dets_nu_times[index] = nu_time
+            # check whether an experiment is revoking its data:
+            if self.t[index] != burst_time:
+                if self.t[index] == None:
+                    if self.valid[index]:
+                        self.valid[index] = False
+                        data['action'] = 'revoke'
+                        return data
+                else:
+                    self.valid[index] = True
+            return False
+
+        def compute_dts(self, data):
+            '''
+            Compute time differences between detector pairs
+            '''
+            # compute and store dts in a 2d list:
+            DeltaTsMatrix = [[0 for j in range(sum(self.valid))] for i in range(sum(self.valid))]
```

Comment:
Do we need all time differences? I think it's actually redundant. For simple time differences between the neturino_time values, then we just need the time differences between, say, each detector and the "best" detector (i.e., smallest uncertainty).  So if we had N detectors, we'd have N-1 time differences.

Time differences from other methods (lightcurve matching, FT's, etc.) can be added, or can replace the values calculated above.  The covariance may get messed up a bit, since there will be some redundant information, but I'm not sure how large an effect that would be.

Reply:
How does one exactly decide which detector has the smallest uncertainty? The best dt should not be the biggest one (i.e. the bigger the dt the smaller the pointing area)? i.e. probably the dt coming from the detectors pairs with the biggest distance (meters) between each other would be the best? I was thinking for example that the dt from the pair (SNOP, HALO) will give us a less efficient pointing than (KM3, KL)

Reply:
One can compare the standard deviations from the data file.

It's true that the time difference between two very close detectors won't count for much; it may be better to calculate the dt of each to a more distant detector.  It's still worth using both, though, because each will have independent data.

So this might argue for calculating more time differences than just every detector minus the one with the smallest sigma.  As long as every detector gets used at least once, it should work.

Reply:
I tried to feed the DiffPointing and if memory does not fail me, it gave good results. 

I must be honest that I do not know starting from which distance comparing two detectors data will be relevant (it also will depend on the direction of the incoming neutrino flux)

The DiffPointing is already reading sigma, bias and variance from the data file so I was not thinking to add them here

Reply:
To get the simplest thing done first:  choose the experiment with the smallest sigma (either from the table; this value may be overridden if the experiment quotes another value in the payload), calculate all differences relative to that one.